### PR TITLE
spack environment (3): expose a subset of packages to a user

### PR DESCRIPTION
**Pull request by scheibelp**
This is the same as https://github.com/spack/spack/pull/7843 but with factored commits. The attribution is incorrect on them (I need to use https://help.github.com/articles/creating-a-commit-with-multiple-authors/ to add @citibeth as a coauthor on most of these).

@tgamblin ready for review. In short this PR:

* Adds the `spack env` command: This maintains an environment object which tracks a set of specs, a spack configuration, and a repository. The user 
  * Adds documentation which explains the usage of this command
  * Adds unit tests
  * Renames the old `spack env` command to `spack build-env`
* Refactors several common arguments in Spack commands so that e.g. `spack install` command options apply to `spack env foo install`
* Addresses some issues with `Spec.to_node_dict` to preserve patches/concreteness, so that a spec is concrete & complete after converting to/from JSON

Unfortunately since this replaces the `cmd/env.py` file, the github diff presents this as if I had edited the file from it's original state (vs. actually I renamed `env.py` and started from scratch). That is to say: it would be easier to just read `cmd/env.py` in an editor since all of its content is new.

### CODE REVIEW

**Comment by citibeth** [comment]
To see how I'm "going to town" with environments, look at:

https://github.com/citibeth/spack/tree/efischer/giss/var/spack/environments

(You need to switch to branch _efischer/giss_)

**Comment by citibeth** [comment]
I am more concerned that this PR gets merged, than I get attribution for it.  Once this PR is merged, I will be actively promoting the Spack Setup PR; and once that is done, I will seek to re-do the environment API so it's more integrated into all Spack commands.  @scheibelp did the heavy lifting here, with me testing and tucking in around the edges.

**Comment by scheibelp** [comment]
> I am more concerned that this PR gets merged, than I get attribution for it.

@citibeth FYI the "fixing attribution" part of this should not consume a significant amount of time. @tgamblin can let me know if it looks good and then I can rebase it and merge myself (just editing the commit messages).

File: lib/spack/docs/environments.rst [hunk 1]
```diff
@@ -0,0 +1,317 @@
+.. _environments:
+
+Environments
+============
+
+An environment is used to group together a set of specs for the
+purpose of building, rebuilding and deploying in a coherent fashion.
+Environments provide a number of advantages over the the a la carte
```
Comment (SteVwonder) [inline lib/spack/docs/environments.rst]: `the the` -> `the`

File: lib/spack/docs/environments.rst [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+   different versions of Spack!
+#. Since re-concretization only occurs at controlled times, the
+   Environment equivalent of many Spack commands are far faster.  For
+   example, ``spack env <env> list -Ilr`` vs. ``spack spec``.
+#. Environments allow several specs to be built at once; a more robust
+   solution than ad-hoc scripts making multiple calls to ``spack
+   install``.
+#. An Environment that is built as a whole can be loaded as a whole.
+   Spack can generate a script of ``module load`` commands that load
+   the appropraite environment.  And that script will work, without
```
Comment (SteVwonder) [inline lib/spack/docs/environments.rst]: `appropraite` -> `appropriate`

File: lib/spack/docs/environments.rst [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+   running Spack, as long as the underlying modules remain installed.
+   A fast-running ``spack env <env> load`` command would also be
+   possible.
+
+Other packaging systems also provide environments that are similar in
+some ways to Spack environments; for example, `Conda environments
+<https://conda.io/docs/user-guide/tasks/manage-environments.html>`_ or
+`Python Virtual Environments
+<https://docs.python.org/3/tutorial/venv.html>`_.  Spack environments
+some distinctive features:
```
Comment (SteVwonder) [inline lib/spack/docs/environments.rst]: Spack environments **provide** some distinctive features

File: lib/spack/docs/environments.rst [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+#. A spec installed "in" an environment is no different form the same
```
Comment (SteVwonder) [inline lib/spack/docs/environments.rst]: `form` -> `from`

File: lib/spack/docs/environments.rst [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+   Modules that occur in earlier specs listed in an environment take
+   precedence over modules that occur later.
+
+Using Environments
+------------------
+
+Here we follow a typical use case of creating, concretizing,
+installing and loading an environment.
+
+.. note::
```
Comment (SteVwonder) [inline lib/spack/docs/environments.rst]: This doesn't seem to be rendering correctly on GitHub.  I'm not familiar with restructured text, so I don't know if that is just GitHub's renderer or if the newlines need to be removed.

File: lib/spack/docs/environments.rst [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+  part of the environment.
+
+  .. note::
+
+     All environments are stored in the ``var/spack/environments`` folder.
+
+  The following files may be added to this directory by the user or
+  Spack:
+
+  * ``env.yaml``: Addition environment specification and configuration
```
Comment (SteVwonder) [inline lib/spack/docs/environments.rst]: `Addition` -> `Additional`

**Comment by scheibelp** [comment]
I have rebased this to sync with develop. I plan on looking more into https://github.com/spack/spack/pull/8231#pullrequestreview-139714571 tomorrow (thanks @SteVwonder!)

**Comment by tgamblin** [comment]
@scheibelp: sorry for not pushing earlier -- I was waiting to do this transactionally.  I pushed my changes, which:
1. fix some bugs that came up after the rebase (I think there were 3-ish bugs in the modules stuff after the initial rebase onto `develop` -- you had the same ones)
2. fix flake8 issues (from pep8-naming)
3. add @SteVwonder's feedback.

This also:

- [x] moves most of `spack.environment` to `spack.util.environment`
- [x] moves the core of spack environments to `spack.environment` instead of keeping it all in the command
- [x] makes the `cmd/spack/env.py` test use fixtures/`pytest
- [x] refactors `cmd/env.py` so that parsers are closer to the corresponding code

I'll post more in a follow up, but the above is a setup for some refactors I want before we put this in and allow people to rely on the API.  Most of them are in line with what @citibeth has requested.  I think this is going to look really good!

**Comment by tgamblin** [comment]
Superseded by #9612

**Pull request closed without merging**